src/Analysis/Codelyzer.Analysis.CSharp/Handlers/ClassDeclarationHandler.cs

```diff
@@ -1,42 +1,77 @@
 using System.Linq;
 
 namespace Codelyzer.Analysis.CSharp.Handlers
 {
     public class ClassDeclarationHandler : UstNodeHandler
     {
+        // Key: class full identifier, value: ClassDeclaration.
+        private Dictionary<string, ClassDeclaration> _classDeclarationCache = new();
         private ClassDeclaration ClassDeclaration { get => (ClassDeclaration)UstNode; }
 
         public ClassDeclarationHandler(CodeContext context,
             ClassDeclarationSyntax syntaxNode)
             : base(context, syntaxNode, new ClassDeclaration())
         {
             var classSymbol = SemanticHelper.GetDeclaredSymbol(syntaxNode, SemanticModel, OriginalSemanticModel);
-
-            ClassDeclaration.Identifier = syntaxNode.Identifier.ToString();
-            ClassDeclaration.Modifiers = syntaxNode.Modifiers.ToString();
-
             if (classSymbol != null)
             {
-                if(classSymbol.BaseType != null)
-                {
-                    ClassDeclaration.BaseType = classSymbol.BaseType.ToString();               
-                    ClassDeclaration.BaseTypeOriginalDefinition = GetBaseTypOriginalDefinition(classSymbol);
-                    ClassDeclaration.Reference.Namespace = GetNamespace(classSymbol);
-                    ClassDeclaration.Reference.Assembly = GetAssembly(classSymbol);
-                    ClassDeclaration.Reference.Version = GetAssemblyVersion(classSymbol);
-                    ClassDeclaration.Reference.AssemblySymbol = classSymbol.ContainingAssembly;
-                    ClassDeclaration.FullIdentifier = classSymbol.OriginalDefinition.ToString();
-                }
-                
-                if(classSymbol.Interfaces != null)
+                var fullIdentifier = GetFullIdentifier(classSymbol);
+                if (_classDeclarationCache.ContainsKey(fullIdentifier))
                 {
-                    ClassDeclaration.BaseList = classSymbol.Interfaces.Select(x => x.ToString())?.ToList();
+                    UstNode = _classDeclarationCache[fullIdentifier];
+                    return;
                 }
+                Set(ClassDeclaration, classSymbol);
+                _classDeclarationCache.Add(fullIdentifier, ClassDeclaration);
+            }
+        }
+
+        private void Set(ClassDeclaration classDeclaration, INamedTypeSymbol classSymbol)
+        {
+            classDeclaration.Identifier = classSymbol.Name;
+            classDeclaration.FullIdentifier = classSymbol.OriginalDefinition.ToString();
+            classDeclaration.Reference.Namespace = GetNamespace(classSymbol);
+            classDeclaration.Reference.Assembly = GetAssembly(classSymbol);
+            classDeclaration.Reference.Version = GetAssemblyVersion(classSymbol);
+            classDeclaration.Reference.AssemblySymbol = classSymbol.ContainingAssembly;
+
+            var syntaxNode = (ClassDeclarationSyntax)classSymbol.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax();
```

Comment:
What if it's an interface?

Reply:
Are you talking about the `classSymbol`? It will always represents a class either the current class or its base class.